archunit/src/main/java/com/tngtech/archunit/library/freeze/ViolationStoreFactory.java

```diff
@@ -0,0 +1,225 @@
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+class ViolationStoreFactory {
+    static final String FREEZE_STORE_PROPERTY = "freeze.store";
+
+    static ViolationStore create() {
+        return ArchConfiguration.get().containsProperty(FREEZE_STORE_PROPERTY)
+                ? createInstance(ArchConfiguration.get().getProperty(FREEZE_STORE_PROPERTY))
+                : new TextFileBasedViolationStore();
+    }
+
+    @MayResolveTypesViaReflection(reason = "This is not part of the import process")
+    private static ViolationStore createInstance(String violationStoreClassName) {
+        try {
+            return (ViolationStore) newInstanceOf(Class.forName(violationStoreClassName));
+        } catch (Exception e) {
+            String message = String.format("Could not instantiate %s of configured type '%s=%s'",
+                    ViolationStore.class.getSimpleName(), FREEZE_STORE_PROPERTY, violationStoreClassName);
+            throw new StoreInitializationFailedException(message, e);
+        }
+    }
+
+    @VisibleForTesting
+    static class TextFileBasedViolationStore implements ViolationStore {
+        private static final Logger log = LoggerFactory.getLogger(TextFileBasedViolationStore.class);
+
+        private static final Pattern UNESCAPED_LINE_BREAK_PATTERN = Pattern.compile("(?<!\\\\)\n");
+        private File storeFolder;
+        private FileSyncedProperties storedRules;
+
+        @Override
+        public void initialize(Properties properties) {
+            String path = properties.getProperty("default.path", "archunit_store");
+            storeFolder = new File(path);
+            ensureExistence(storeFolder);
+            File storedRulesFile = new File(storeFolder, "stored.rules");
+            log.info("Initializing {} at {}", TextFileBasedViolationStore.class.getSimpleName(), storedRulesFile.getAbsolutePath());
+            storedRules = new FileSyncedProperties(storedRulesFile);
+            checkInitialization(storedRules.initializationSuccessful(), "Cannot create rule store at %s", storedRulesFile.getAbsolutePath());
+        }
+
+        private void ensureExistence(File folder) {
+            checkState(folder.exists() && folder.isDirectory() || folder.mkdirs(), "Cannot create folder %s", folder.getAbsolutePath());
+        }
+
+        private void checkInitialization(boolean initializationSuccessful, String message, Object... args) {
+            if (!initializationSuccessful) {
+                throw new StoreInitializationFailedException(String.format(message, args));
+            }
+        }
+
+        @Override
+        public boolean contains(ArchRule rule) {
+            return storedRules.containsKey(rule.getDescription());
+        }
+
+        @Override
+        public void save(ArchRule rule, List<String> violations) {
+            log.debug("Storing evaluated rule '{}' with violations: {}", rule.getDescription(), violations);
+
+            UUID ruleId = ensureRuleId(rule);
```

Comment:
I don't know if I want to encourage people to tinker with the internal structure of the text file based `ViolationStore` :wink: But I also don't have strong feelings about this...
I any case, I don't understand why you've changed the original debug log that a rule is already stored to an error log? I don't think it's an error case if the rule is already present within the stored properties, because that should be a pretty standard case when existing violations are adjusted or am I mistaken?

Reply:
I've switched the log level back to debug, because within the unit test the normal flow of updating violations would now log errors, so I'm pretty sure debug is okay :wink:

Reply:
Thanks for catching this! I maybe shouldn't try to make contributions late at night before going on vacation...
I didn't intend to commit the changed log level, but only used it to manually (integration-)test the log statement without changing the log configuration in my project. Sorry for not looking at the diff carefully enough!

Reply:
No problem, I just thought I've missed something here :wink: No harm done